### ISSUE TITLE
Implement email verification for users 

### ISSUE THREAD

**Issue by Radhikadua123** [publiclab/plots2#3798]
Right now we don't have any kind of email verification. User can register using any email, user can change his/her email and then just use the account. He/she won't have to confirm their emails to do any actions. This can exploited by users for spamming too.

So, this is parent issue describing all the smaller bugs which needs to be fixed to successfully add email verification system. This implementation corresponds to my second approach here: https://github.com/publiclab/plots2/issues/3522#issuecomment-433109387

- [ ] Implement helper function to encrypt/decrypt data
- [ ] Implement helper function to generate and validate unique token for user
- [ ] Add is_verified to user model with default value `False`
- [ ] Send verification email to the user on signup
- [ ] Flash message and button to send email again if email not verified
- [ ] Implement helper function has_email_verified  
- [ ] Disable write operations on pages if user not verified

Optional:
- [ ] Add redis database to docker and connect it with application
- [ ] Implement expiry for the validation link

@jywarren Should I create separate issues for each of them or this much breaking is enough and easy to understand ?

**Comment by jywarren** [comment]
this is amazingly well broken up! Where did you learn this workflow -- it's
one we've tried to be good at, but it's tough! See our posts at
https://publiclab.org/software-outreach

>
I think you can do them all yourself in this issue if you want. But if
you'd like to prepare one or two for others to get involved in, breaking
them out into well-documented distinct issues can help. But it's up to you!

Thank you!

**Comment by Radhikadua123** [comment]
> this is amazingly well broken up! Where did you learn this workflow -- it's one we've tried to be good at, but it's tough! 

Thanks a lot! @jywarren ^_^ 
Actually I only know general web development stuff and I thought lot about the way it should be implemented and then tried to split it into smaller tasks. But I'm quite less aware of Ruby on Rails internals. I'm continuously trying to learn more about it from the code base. 

> See our posts at https://publiclab.org/software-outreach

Sure, I'll check it out.

> I think you can do them all yourself in this issue if you want. But if you'd like to prepare one or two for others to get involved in, breaking them out into well-documented distinct issues can help.

Yeah, splitting task into well-documented distinct issues on which others can work easily sounds better idea to me. By this, more people can get involved in the project.

**Comment by jywarren** [comment]
👍🏽👍🏽👍🏽

**Comment by shubhangi-chauhan** [comment]
Could I take this up? :smiley:

**Comment by Radhikadua123** [comment]
@shubhangi-chauhan Sure, feel free to work on it. I'll add new well documented issues and link them here today.

**Comment by shubhscoder** [comment]
I would like to work on this, if anyone has'nt taken this issue. :)

**Comment by Radhikadua123** [comment]
I have split up this issue into separate issues. Everyone please feel free to take subparts of this issue!

**Comment by faithngetich** [comment]
@Radhikadua123 can they all be worked on by different people at a different time or one issue does not block the other?

**Comment by Radhikadua123** [comment]
Hi @faithngetich,

Actually lot of them depend on the previous tasks. But initial tasks like ` Implement helper function to generate and validate unique token for user ` and `Add is_verified to user model with default value False` can be done in parallel.

**Comment by faithngetich** [comment]
Thanks @Radhikadua123 

I see they've all been worked on.

**Comment by SidharthBansal** [comment]
@Radhikadua123 really you did a great job in breaking the stuff into chunks.

**Comment by Radhikadua123** [comment]
Thanks a lot! @SidharthBansal

**Comment by Radhikadua123** [comment]
@faithngetich May be you can work on something else ? 

In case you need help in search of appropriate task, I can try finding one for you.

**Comment by SidharthBansal** [comment]
Thanks @Radhikadua123
